lib/AdobeState.js

```diff
@@ -72,39 +72,44 @@ async function _wrap (promise, params) {
   try {
     response = await promise
     logger.debug('response', response)
-    // reuse code in exception handler, for any other network exceptions
-    if (!response.ok) {
-      // no exception on 404
-      if (response.status === 404) {
-        return null
-      } else {
-        const e = new Error(response.statusText)
-        e.status = response.status
-        e.internal = response
-        throw e
-      }
-    }
   } catch (e) {
-    const status = e.status || e.code
-    const copyParams = cloneDeep(params)
-    logger.debug(`got internal error with status ${status}: ${e.message} `)
-    switch (status) {
-      case 401:
-        return logAndThrow(new codes.ERROR_UNAUTHORIZED({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
-      case 403:
-        return logAndThrow(new codes.ERROR_BAD_CREDENTIALS({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
-      case 413:
-        return logAndThrow(new codes.ERROR_PAYLOAD_TOO_LARGE({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
-      case 429:
-        return logAndThrow(new codes.ERROR_REQUEST_RATE_TOO_HIGH({ sdkDetails: copyParams }))
-      default:
-        // NOTE: we should throw a different error if its not a response error
-        return logAndThrow(new codes.ERROR_INTERNAL({ messageValues: [`unexpected response from provider with status: ${status}`], sdkDetails: { ...cloneDeep(params), _internal: e.internal || e.message } }))
-    }
+    logAndThrow(e)
   }
+
+  handleResponse(response, params)
   return response
 }
 
+/**
+ * Handle a network response.
+ *
+ * @param {Response} response a fetch Response
+ * @param {object} params the params to the network call
+ * @returns {void}
+ */
+function handleResponse (response, params) {
+  if (response.ok) {
+    return
+  }
+
+  const copyParams = cloneDeep(params)
+  switch (response.status) {
+    case 404:
+      // do nothing, no exception on 404
+      break
+    case 401:
+      return logAndThrow(new codes.ERROR_UNAUTHORIZED({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
+    case 403:
+      return logAndThrow(new codes.ERROR_BAD_CREDENTIALS({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
+    case 413:
+      return logAndThrow(new codes.ERROR_PAYLOAD_TOO_LARGE({ messageValues: ['underlying DB provider'], sdkDetails: copyParams }))
+    case 429:
+      return logAndThrow(new codes.ERROR_REQUEST_RATE_TOO_HIGH({ sdkDetails: copyParams }))
+    default: // 500 errors
+      return logAndThrow(new codes.ERROR_INTERNAL({ messageValues: [`unexpected response from provider with status: ${response.status}`], sdkDetails: copyParams }))
```

Comment:
should we include the response body for unexpected errors?

Reply:
i'll add it

Reply:
added via https://github.com/adobe/aio-lib-state/pull/144/commits/72e7b45d5c5862e292db4864b1eff6c517a72f8e